File: widgetssdk/src/main/java/com/glia/widgets/chat/domain/CustomCardShouldShowUseCase.java
package com.glia.widgets.chat.domain;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.glia.androidsdk.chat.ChatMessage;
import com.glia.androidsdk.chat.SingleChoiceAttachment;
import com.glia.widgets.chat.adapter.CustomCardAdapter;

/**
 * @hide
 */
public class CustomCardShouldShowUseCase {
    @Nullable
    private final CustomCardAdapter adapter;

    public CustomCardShouldShowUseCase(@Nullable CustomCardAdapter adapter) {
        this.adapter = adapter;
    }

    public boolean execute(@NonNull ChatMessage message, int viewType, boolean shouldApplySelectedOption) {
        if (shouldApplySelectedOption && !hasSelectedOption(message)) {
            return true;
        }
        if (adapter == null || message.getMetadata() == null) {
            return false;
        }
        return adapter.shouldShowCard(message, viewType);
    }

    private boolean hasSelectedOption(@NonNull ChatMessage message) {
        if (message.getAttachment() == null
                || !(message.getAttachment() instanceof SingleChoiceAttachment)) {
            return false;
        }
        SingleChoiceAttachment singleChoiceAttachment = (SingleChoiceAttachment) message.getAttachment();
        String selectedOption = singleChoiceAttachment.getSelectedOption();
        return selectedOption != null && !selectedOption.isEmpty();
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/chat/domain/CustomCardShouldShowUseCase.kt
package com.glia.widgets.chat.domain

import com.glia.androidsdk.chat.ChatMessage
import com.glia.androidsdk.chat.SingleChoiceAttachment
import com.glia.widgets.chat.adapter.CustomCardAdapter

internal class CustomCardShouldShowUseCase(private val adapter: CustomCardAdapter?) {
    fun execute(message: ChatMessage, viewType: Int, shouldApplySelectedOption: Boolean): Boolean {
        if (shouldApplySelectedOption && !hasSelectedOption(message)) {
            return true
        }
        if (adapter == null || message.metadata == null) {
            return false
        }
        return adapter.shouldShowCard(message, viewType)
    }

    private fun hasSelectedOption(message: ChatMessage): Boolean {
        if (message.attachment == null || message.attachment !is SingleChoiceAttachment) return false
        val singleChoiceAttachment = message.attachment as SingleChoiceAttachment
        val selectedOption = singleChoiceAttachment.selectedOption
        return selectedOption != null && selectedOption.isNotEmpty()
    }
}
